src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Handlers/SourceItemHandler.cs

```diff
@@ -80,8 +80,12 @@ public void Handle(CommandLineArguments added, CommandLineArguments removed, IWo
 
             if (diff.AddedItems.Count > 0 || diff.RenamedItems.Count > 0 || diff.ChangedItems.Count > 0)
             {
-                // Make sure the tree matches the same version of the evaluation that we're handling
-                IProjectTreeServiceState treeState = await PublishTreeAsync(e).ConfigureAwait(true); // TODO: https://github.com/dotnet/roslyn-project-system/issues/353
+                // Make sure the tree matches the same version of the evaluation that we're handling for active project context.
+                // If we are dealing with non- active project context, then just use the latest published tree.
+                // TODO: https://github.com/dotnet/roslyn-project-system/issues/353
+                IProjectTreeServiceState treeState = isActiveContext ?
```

Comment:
I think this is a good workaround for now, but I really think the right term fix here is to give up on publishing the tree to get the folder names. Instead, I think we should just look calculate it ourselves directly based on the item coming in. Avoids this whole madness of coordinating with the tree just to get the folders. I did this to avoid the Linked item calculation - but looks like this has caused more issues than it saved.

For example, if an item is conditioned out of the current active "TFM", it will never be found in the tree and then when it becomes "active" (perhaps the first TFM was deleted), it will then have the wrong values for "folderNames".

Can you also put the TODO, back on the ConfigureAwait()?